### PR TITLE
fix: catch exception in udf and udsink

### LINKED ISSUE #16: wrap the handlers using try/catch
Today we invoke the handlers without any exception handling. The exceptions are caught in the python gRPC code and are propagated up to the numaflow gRPC client. We should catch the exception before it reaches the gRPC client.

We should catch the exception and log in the UDF log, so the user can easily differentiate between platform error v/s user error.

eg of one such occurrence,

`rpc error: code = Unknown desc = Unexpected <class 'TypeError'>: 'module' object is not callable"`

This was a user code error but the user couldn't differentiate between a platform error v/s user code error.

```json
{"level":"error","ts":1665182691.4869258,"logger":"numaflow.udf-processor","caller":"forward/forward.go:387","msg":"UDF.Apply error","vertex":"numalogic-simple-pipeline-input","protocol":"uds-grpc-udf","error":"gRPC client.MapFn failed, failed to execute c.grpcClt.MapFn(): rpc error: code = Unknown desc = Unexpected <class 'TypeError'>: 'module' object is not callable","stacktrace":"github.com/numaproj/numaflow/pkg/isb/forward.(*InterStepDataForward).applyUDF\n\t/home/runner/work/numaflow/numaflow/pkg/isb/forward/forward.go:387\ngithub.com/numaproj/numaflow/pkg/isb/forward.(*InterStepDataForward).concurrentApplyUDF\n\t/home/runner/work/numaflow/numaflow/pkg/isb/forward/forward.go:372\ngithub.com/numaproj/numaflow/pkg/isb/forward.(*InterStepDataForward).forwardAChunk.func1\n\t/home/runner/work/numaflow/numaflow/pkg/isb/forward/forward.go:210"}
```

### PR DESCRIPTION
**Pull request by ab93**
Closes #16

### CODE REVIEW

**Review by chromevoid — APPROVED** [review]

File: pynumaflow/function/server.py [hunk 1]
```diff
@@ -64,14 +62,18 @@ def MapFn(
...
+            msgs = self.__map_handler(
+                key,
+                Datum(
+                    value=request.value,
+                    event_time=request.event_time.event_time.ToDatetime(),
+                    watermark=request.watermark.watermark.ToDatetime(),
+                ),
+            )
+        except Exception as err:
+            _LOGGER.exception("UDFError: %r", err)
```
Comment (vigith) [inline pynumaflow/function/server.py:75]:
```suggestion
            _LOGGER.critical("UDFError, dropping message on the floor: %r", err)
```

File: pynumaflow/function/server.py [hunk 2]
```diff
@@ -64,14 +62,18 @@ def MapFn(
...
+            msgs = Messages.as_forward_all(None)
```
Comment (vigith) [inline pynumaflow/function/server.py:76]:
```suggestion
            # a neat hack to drop
            msgs = Messages.as_forward_all(None)
```

File: pynumaflow/sink/server.py [hunk 3]
```diff
@@ -57,11 +56,20 @@ def SinkFn(
         The pascal case function name comes from the generated udsink_pb2_grpc.py file.
         """
 
-        msgs = self.__sink_handler(request.elements)
+        try:
```
Comment (vigith) [inline pynumaflow/sink/server.py:59]:
```suggestion
        # if there is an exception, we will mark all the responses as a failure
        try:
```

File: pynumaflow/sink/server.py [hunk 4]
```diff
@@ -57,11 +56,20 @@ def SinkFn(
...
+            rspns = Responses()
+            for _datum in request.elements:
+                rspns.append(Response.as_failure(_datum.id, err_msg))
 
         responses = []
-        for msg in msgs.items():
-            responses.append(udsink_pb2.Response(id=msg.id, success=msg.success, err_msg=msg.err))
+        for rspn in rspns.items():
+            responses.append(
+                udsink_pb2.Response(id=rspn.id, success=rspn.success, err_msg=rspn.err)
```
Comment (ab93) [inline pynumaflow/sink/server.py:71]: @vigith this part is good, actually. We are creating a new list `responses` from `rspns` regardless of the exception or not.

File: pynumaflow/sink/server.py [hunk 5]
```diff
@@ -55,12 +55,12 @@ def SinkFn(
         Applies a sink function to a list of datum elements.
         The pascal case function name comes from the generated udsink_pb2_grpc.py file.
         """
-
+        # if there is an exception, we will mark all the responses as a failure
         try:
             rspns = self.__sink_handler(request.elements)
         except Exception as err:
```
Comment (vigith) [inline pynumaflow/sink/server.py:61]: any reason you didn't try try-catch-else?

**Review by vigith — APPROVED** [review]

File: pynumaflow/sink/server.py (hunk 5 above)
Comment (ab93) [reply pynumaflow/sink/server.py:61]: > @vigith this part is good, actually. We are creating a new list `responses` from `rspns` regardless of the exception or not.

Added the reason in the previous comment! :)

**Pull request merged**